### PR TITLE
Documentation eventhandler

### PR DESCRIPTION
**Pull request by MaGaMeGa**

### CODE REVIEW

**Review by borditamas — APPROVED** [review]

File: documentation/eventhandler/Publish.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+```
+
+| __PublishRequest__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `eventType` | Type of event. | mandatory | max. length = 255 |
+| `metaData` |  The "key - value" pairs for event filtering. | optional | max.length = 65535 |
+| `payload` | String representation of the event. | mandatory | specifyed at system deployment time |
```
Comment (tsvetlin) [inline documentation/eventhandler/Publish.md]: specified

File: documentation/eventhandler/Subscribe.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  }
+}
+```
+| __SubscriptionRequest__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `eventType` | Type of event to subscribe to | mandatory | max. length = 255 |
+| `filterMetaData` | The recievied event have to contain all the "key - value" pairs defined here  | optional | max.length = 65535 |
```
Comment (tsvetlin) [inline documentation/eventhandler/Subscribe.md]: has to contain

File: documentation/eventhandler/Subscribe.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+| `matchMetaData` | A flag to turn on/off metadata filtering | mandatory |  ture or false |
```
Comment (tsvetlin) [inline documentation/eventhandler/Subscribe.md]: true of false

File: documentation/eventhandler/Subscribe.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+| `notifyUri` | Url subpath of the subscriber sytem's notification endpoint | mandatory | max.length = 65535 |
+| `sources` | List of publisher systems | optional (if not difined or empty, all publishers will be able to send requests which are authorized and allowed by the other filtering options )| specifyed at system deployment time |
```
Comment (tsvetlin) [inline documentation/eventhandler/Subscribe.md]: defined, specified

File: documentation/eventhandler/Subscribe.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+| `endDate` | If endDate is defined, the subscriber system will only receive events when the events timestamp is before endDate. | optional ( EndDate must be after the current datetime. If startDate is defined endDate must be after startDate. )|  UTC time in `yyyy-MM-dd`  `HH`:`mm`:`ss` format  |
+| `subscriberSystem` | Details of subscriber system | mandatory | as in system |
+
+| __System__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `systemName` | The name of the system. | mandatory | max. length = 255 |
+| `address` |  Domain name or Ip of the system. | mandatory | max. length = 255 |
```
Comment (tsvetlin) [inline documentation/eventhandler/Subscribe.md]: IP

File: documentation/eventhandler/Subscribe.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+| `authenticationInfo` | Public key of the system. | optional | single line string without the "-----BEGIN PUBLIC KEY-----" prefix  and the "-----END PUBLIC KEY-----" suffix |
+| `port` | The port where the system servs it's services | mandatory | max.length = difined by local cloud operator ( default valid range: 1-65535 ) |
```
Comment (tsvetlin) [inline documentation/eventhandler/Subscribe.md]: provides

Comment (tsvetlin) [reply documentation/eventhandler/Subscribe.md]: defined

File: documentation/eventhandler/Unsubscribe.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+__Unsubscribe query parameters__ are the input :
+`https://eventhandler_ip:unsubscribe_port/eventhandler/unsubscribe?address=`192.168.0.1`&event_type=`EVENT_TYPE_1`&port=`9009`&system_name=`test_consumer
+
+| __Unsubscribe__  query parameters |
+| ------------------------------------------------------- |
+
+| Parameter | Description | Necessity | Format/Limitations |
+| --------- | ----------- | --------- | ----------- |
+| `address` |  Domain name or Ip of the system. | mandatory | max. length = 255 |
```
Comment (tsvetlin) [inline documentation/eventhandler/Unsubscribe.md]: IP

File: documentation/eventhandler/Unsubscribe.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+| `event_type` | Type of event to subscribe to | mandatory | max. length = 255 |
+| `port` | The port where the system servs it's services | mandatory | max.length = difined by local cloud operator ( default valid range: 1-65535 ) |
```
Comment (tsvetlin) [inline documentation/eventhandler/Unsubscribe.md]: provides, defined

File: documentation/eventhandler/use_cases/EH_use_case_1.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+| Main Flow | - The Publisher sends an event to the Event Handler. <br/>- The Event Handler validate: <br/>- - the request format <br/>- - the Event source <br/>- - the Event TimeStamp. <br/>- The Event Handler select all Subscribers by the type of the event<br/>- The Event Handler filters Subscribers by authorization.<br/>- The Event Handler filters Subscribers by timestamp.<br/>- The Event Handler filters Subscribers by metadata.<br/>- The Event Handler put the event with the filtered subscribers list into the publishing queue.<br/>-  The Event Handler returns "ok" to publisher.<br/>- PublishingQueueWatcherTask senses new item in the publishing queue.<br/>- PublishingQueueWatcherTask, based on the number of subscribers, pass the event with the filtered subscriber list ether to the expressExecutor ( which has a thread pool initialized at startup time ) or creates a new publishRequestExecutor (with its own thread pool ). <br/>- The used Executor,  for each Subscriber, execute the PublishEventTask, where it removes the publisher from the event and send the result to the subscriber.
```
Comment (tsvetlin) [inline documentation/eventhandler/use_cases/EH_use_case_1.md]: passes, 
ether? together?
there are two spaces before "for each Subscriber"

File: documentation/eventhandler/use_cases/EH_use_case_2.md [hunk 10]
```diff
@@ -0,0 +1,10 @@
+Use case 2 : *Register Subscription*
+
+| Name | Description |
+| ---- | --------- |
+| ID | Subscription |
+| Brief Description | The Even Handler is tasked to register a subscription |
+| Primary Actors | Subscriber, Event Handler, Authorization Core Service |
+| Preconditions | Subscriber has to be registered in Service Registry  |
+| Main Flow | - The Subscriber sends a SubscriptionRequest  to the Event Handler. <br/>- The Event Handler denial any request on it's Subscription endpoint if it is not from the system defined as the SubscriberSystem in the SubscriptionRequest<br/>- The Event Handler validate: <br/>- - the request format <br/>- - the request parameters validity. <br/>-  The Event Handler query the Authorization Core Service for the authorized providers of the subscriber as a consumer.<br/>- The Event Handler registers the Subscription and the SubscriptionEntryPublisherConnections. |
```
Comment (tsvetlin) [inline documentation/eventhandler/use_cases/EH_use_case_2.md]: Authorization Core Service -> Authorization Core System
two spaces,
denial -> denies,
The Event Handler validate: -> validates,
The Event Handler query -> queries,

File: documentation/eventhandler/use_cases/EH_use_case_3.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+Use case 3 : *Unregister Subscription*
+
+| Name | Description |
+| ---- | --------- |
+| ID | Unsubscribe |
+| Brief Description | The Even Handler is tasked to unregister a subscription |
+| Primary Actors | Subscriber, Event Handler|
+| Preconditions |  Subscriber has to be registered in Service Registry  |
+| Main Flow | - The Subscriber sends an id and a SubscriptionRequest  to the Event Handler's unregister endpoint. <br/>- The Event Handler denial any request on it's Unsubscription endpoint if it is not from the system defined as the SubscriberSystem in the SubscriptionRequest. <br/>- The Event Handler validate: <br/>- - the request format <br/>- - the request parameters validity.<br/>- The Event Handler unregisters the SubscriptionEntryPublisherConnections and the Subscription. |
```
Comment (tsvetlin) [inline documentation/eventhandler/use_cases/EH_use_case_3.md]: denies, validates

File: documentation/eventhandler/use_cases/EH_use_case_4.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
+Use case 4: *Publish Authorization Update Event*
+
+| Name | Description |
+| ---- | --------- |
+| ID | Publish-Auth-Update-Event |
+| Brief Description | The Even Handler is tasked to update the allowed providers for the given Subscriber |
```
Comment (tsvetlin) [inline documentation/eventhandler/use_cases/EH_use_case_4.md]: Event Handler

File: documentation/eventhandler/use_cases/EH_use_case_4.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+| Primary Actors | Authorization Core Service, Event Handler |
+| Preconditions | -   Subscriber has to be registered to event handler  |
```
Comment (tsvetlin) [inline documentation/eventhandler/use_cases/EH_use_case_4.md]: Event Handler

File: documentation/eventhandler/use_cases/EH_use_case_4.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+| Main Flow | - The Authorization Core Service sends an AuthorizationUpdateEvent to the Event Handler. <br/>- The Event Handler denial any request on it's Publish-Auth-Update endpoint if it is not from the Authorization Core Service<br/>- The Event Handler validate: <br/>- - the request format <br/>- - the Event source <br/>- - the Event TimeStamp. <br/>- - the AuthorizationUpdateEvent Type. <br/>- - the AuthorizationUpdateEvent Payload. <br/>- The Event Handler select the involved Subscriber by the request payload.<br/>- The Event Handler query the Authorization Core Service for the authorized providers of the subscriber as a consumer.<br/>- The Event Handler updates the authorization status of the subscribers SubscriptionEntryPublisherConnections. |
```
Comment (tsvetlin) [inline documentation/eventhandler/use_cases/EH_use_case_4.md]: denies, Core System, selects, queries

File: documentation/eventhandler/use_cases/EH_use_case_1.md (hunk 9 above)
Comment (MaGaMeGa) [reply documentation/eventhandler/use_cases/EH_use_case_1.md]: fixed

File: README.md [hunk 15]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+  }
+}
+```
+| __SubscriptionRequest__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `eventType` | Type of event to subscribe to | mandatory | max. length = 255 |
+| `filterMetaData` | The recievied event have to contain all the "key - value" pairs defined here  | optional | max.length = 65535 |
```
Comment (tsvetlin) [inline README.md]: has to contain

File: README.md [hunk 16]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `matchMetaData` | A flag to turn on/off metadata filtering | mandatory |  ture or false |
+| `notifyUri` | Url subpath of the subscriber sytem's notification endpoint | mandatory | max.length = 65535 |
+| `sources` | List of publisher systems | optional (if not difined or empty, all publishers will be able to send requests which are authorized and allowed by the other filtering options )| not defined |
```
Comment (tsvetlin) [inline README.md]: defined

File: README.md [hunk 17]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+```
+| __SubscriptionRequest__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `eventType` | Type of event to subscribe to | mandatory | max. length = 255 |
+| `filterMetaData` | The recievied event have to contain all the "key - value" pairs defined here  | optional | max.length = 65535 |
+| `matchMetaData` | A flag to turn on/off metadata filtering | mandatory |  ture or false |
+| `notifyUri` | Url subpath of the subscriber sytem's notification endpoint | mandatory | max.length = 65535 |
```
Comment (tsvetlin) [inline README.md]: systems

File: README.md [hunk 18]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `endDate` | If endDate is defined, the subscriber system will only receive events when the events timestamp is before endDate. | optional ( EndDate must be after the current datetime. If startDate is defined endDate must be after startDate. )|  UTC time in `yyyy-MM-dd`  `HH`:`mm`:`ss` format  |
+| `subscriberSystem` | Details of subscriber system | mandatory | as in system |
+
+| __System__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `systemName` | The name of the system. | mandatory | max. length = 255 |
+| `address` |  Domain name or Ip of the system. | mandatory | max. length = 255 |
```
Comment (tsvetlin) [inline README.md]: IP

File: README.md [hunk 19]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `authenticationInfo` | Public key of the system. | optional | single line string without the "-----BEGIN PUBLIC KEY-----" prefix  and the "-----END PUBLIC KEY-----" suffix |
+| `port` | The port where the system servs it's services | mandatory | max.length = difined by local cloud operator ( default valid range: 1-65535 ) |
```
Comment (tsvetlin) [inline README.md]: serves, defined

File: README.md [hunk 20]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+
+__Unsubscribe query parameters__ are the input :
+`https://eventhandler_ip:unsubscribe_port/eventhandler/unsubscribe?address=`192.168.0.1`&event_type=`EVENT_TYPE_1`&port=`9009`&system_name=`test_consumer`
+
+| __Unsubscribe__  query parameters |
+| ------------------------------------------------------- |
+
+| Parameter | Description | Necessity | Format/Limitations |
+| --------- | ----------- | --------- | ----------- |
+| `address` |  Domain name or Ip of the system. | mandatory | max. length = 255 |
```
Comment (tsvetlin) [inline README.md]: IP

File: README.md [hunk 21]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `event_type` | Type of event to subscribe to | mandatory | max. length = 255 |
+| `port` | The port where the system servs it's services | mandatory | max.length = difined by local cloud operator ( default valid range: 1-65535 ) |
```
Comment (tsvetlin) [inline README.md]: serves, defined

File: README.md [hunk 22]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `source` |   Details of the publisher system. | mandatory | as in system |
+| `timestamp` | The time of publishing  | mandatory | UTC time in `yyyy-MM-dd`  `HH`:`mm`:`ss` format |
+
+| __System__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `systemName` | The name of the system. | mandatory | max. length = 255 |
+| `address` |  Domain name or Ip of the system. | mandatory | max. length = 255 |
```
Comment (tsvetlin) [inline README.md]: IP

File: README.md [hunk 23]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `authenticationInfo` | Public key of the system. | optional | single line string without the "-----BEGIN PUBLIC KEY-----" prefix  and the "-----END PUBLIC KEY-----" suffix |
+| `port` | The port where the system servs it's services | mandatory | max.length = difined by local cloud operator ( default valid range: 1-65535 ) |
```
Comment (tsvetlin) [inline README.md]: serves, defined

File: README.md (hunk 17 above)
Comment (rbocsi) [reply README.md]: system's

File: README.md [hunk 24]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+  }
+}
+```
+| __SubscriptionRequest__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `eventType` | Type of event to subscribe to | mandatory | max. length = 255 |
+| `filterMetaData` | The recievied event have to contain all the "key - value" pairs defined here  | optional | max.length = 65535 |
```
Comment (tsvetlin) [inline README.md]: has to

File: README.md [hunk 25]
```diff
@@ -4194,27 +4190,512 @@ __PriorityList__ is the input
...
+| `matchMetaData` | A flag to turn on/off metadata filtering | mandatory |  ture or false |
+| `notifyUri` | Url subpath of the subscriber sytem's notification endpoint | mandatory | max.length = 65535 |
+| `sources` | List of publisher systems | optional (if not difined or empty, all publishers will be able to send requests which are authorized and allowed by the other filtering options )| not defined |
```
Comment (tsvetlin) [inline README.md]: defined

File: documentation/eventhandler/Publish.md [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+| `source` |   Details of the publisher system. | mandatory | as in system |
+| `timestamp` | The time of publishing  | mandatory | UTC time in `yyyy-MM-dd`  `HH`:`mm`:`ss` format |
+
+| __System__  fields |
+| ------------------------------------------------------- |
+
+| Field | Description | Necessity | Format/Limitations |
+| ----- | ----------- | --------- | ----------- |
+| `systemName` | The name of the system. | mandatory | max. length = 255 |
+| `address` |  Domain name or Ip of the system. | mandatory | max. length = 255 |
```
Comment (tsvetlin) [inline documentation/eventhandler/Publish.md]: IP

File: documentation/eventhandler/Publish.md [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+| `authenticationInfo` | Public key of the system. | optional | single line string without the "-----BEGIN PUBLIC KEY-----" prefix  and the "-----END PUBLIC KEY-----" suffix |
+| `port` | The port where the system servs it's services | mandatory | max.length = difined by local cloud operator ( default valid range: 1-65535 ) |
```
Comment (tsvetlin) [inline documentation/eventhandler/Publish.md]: serves

Comment (tsvetlin) [reply documentation/eventhandler/Publish.md]: defined

**Pull request merged**